email and api clients

kashifch opened the pull request:
@benpatterson 

Please take a look whenever you have some time.

We have three different classes for APIs cause it look like different apis hare handled differently, the ecommerce apis could be handled using edx-api-request-client but the for lms apis we need to create a session for logged in user, the enrollment api needs the ACCESS TOKEN of a staff user

kashifch commented:
@benpatterson please take a look when you find some time, also it seems that there is a port issue on jenkins that is not allowing my scripts to access gmail. Since it was working correctly in the past there may be some enhanced security settings or something which is causing the blockage. please do take a look when you have some time

benpatterson commented on regression/pages/common/api_clients.py:
```diff
@@ -0,0 +1,260 @@
+        response:
+    """
+    if response.status_code != 200:
+        raise ApiException(
```
you may want to include response.text as well.

benpatterson replied:
In the case of a non-200 code though, I would expect there to be less text. My main point is to think through what will help someone who is troubleshooting the issue. If you think the text is going to be more noise than it's worth, then that's ok. You can stick with what you have. But pls look at a couple of examples of failures before you decide.

benpatterson commented on regression/pages/common/api_clients.py:
```diff
@@ -0,0 +1,260 @@
+
+    def __call__(self, r):
+        """ Update the request headers. """
+        r.headers['Authorization'] = 'Bearer {}'.format(self.token)
```
I seem to remember that we don't have pylint on this repo? One-letter variables are not the best...could you please re-name?

kashifch replied:
done

benpatterson commented on regression/pages/common/api_clients.py:
```diff
@@ -0,0 +1,260 @@
+        Args:
+            payload:
+        """
+        response = self.ecommerce_api_client.coupons.post(data=payload)
```
Do you need headers in here? Do those come for free with the client?

kashifch replied:
no need for headers when using edxRestApiClient

benpatterson replied:
Ok

benpatterson commented on regression/pages/common/api_clients.py:
```diff
@@ -0,0 +1,260 @@
+        """
+        response = self.ecommerce_api_client.coupons.post(data=payload)
+        if 'coupon_id' not in response.keys():
+            raise ApiException('Coupon not created')
```
It might be more accurate to say "Coupon not returned."

benpatterson commented on regression/pages/common/api_clients.py:
```diff
@@ -0,0 +1,260 @@
+            coupon_id:
+        """
+        response = self.ecommerce_api_client.products(coupon_id).get()
+        if not response:
```
Is this better as a try/except? Would we really get a `None` object if there was a failure in a `get`?

kashifch replied:
Client only returns a dictionary of values and does not return any status code or something so I could not check the 200 or ok status. In case of any errors the client will raise a slumber http exception which i don't think we need to catch, the only eventuality I am covering here is that if client returns an empty list or dictionary. sometimes the coupon ids can exist without any valid data attached with these so in that scenario I would like to raise an exception

benpatterson replied:
Ok I understand. `products().get()` would return `None`?

benpatterson commented on regression/pages/common/api_clients.py:
```diff
@@ -0,0 +1,260 @@
+            coupon_id:
+        """
+        response = self.ecommerce_api_client.coupons(coupon_id).delete()
+        if not response:
```
Same question here....will an exception be raised if something goes wrong on the `.delete()` call? I really don't know...it's likely in the client but I wanted to ask.

kashifch replied:
removing the return here, if anything goes wrong the slumber http exception will be generated

benpatterson commented on regression/pages/common/api_clients.py:
```diff
@@ -0,0 +1,260 @@
+        )
+        for item in items:
+            coupon_codes.append(item['code'])
+        return coupon_codes
```
Will you see repeated codes in the report? Is that something you want to worry about in the list that you return?

kashifch replied:
nope, the codes are unique

benpatterson commented on regression/pages/common/api_clients.py:
```diff
@@ -0,0 +1,260 @@
+        payload = {
+            'email': user_email,
+            'password': user_password,
+            'remember': 'false'
```
What's `remember` do?

kashifch replied:
it is the value for remember me check box below the login

benpatterson replied:
Ok. A comment would help in this case.

benpatterson commented on regression/pages/common/api_clients.py:
```diff
@@ -0,0 +1,260 @@
+        self.session.headers = self._post_headers(
+            response.cookies['csrftoken']
+        )
+        self.login_response = response
```
Do you want this method to handle or return any errors or failures?

kashifch replied:
the check_response function is checking for errors

benpatterson commented on regression/pages/common/api_clients.py:
```diff
@@ -0,0 +1,260 @@
+            course_id:
+            enrollment_action:
+        """
+        change_enrollment_url = self.host + 'change_enrollment'
```
This is going to assume that `self.host` ends with a slash. You may want to handle that if it's not the case.

kashifch replied:
all our base urls end with a slash so it shouldn't be an issue

benpatterson replied:
They end with a slash *now* but someone may make a change later in which they don't. Or will set an environment variable where they don't. So you just want to think through that and assume someone makes a manual mistake somewhere. (If it were me, I would not know if I need a slash or not, so it's a bit arbitrary what I might set locally.)

You need either instructions to end with a slash, or (better) a utility that ensures `self.host` ends with a slash.

kashifch commented:
The activation, password change and bulk purchasing all send email to users (old user and new ones). Like in case of activation I need to have a valid email address for registration and after registration I need to go to that inbox and read the mail user has received and get the activation link from it. So I have created a gmail account for testing and for all new users I use an alias, I use the python imap to connect to gmail inbox.

Previously I was using tempmail api, where I could create a temporary account on tempmail and read the inbox using the api but the problem was  that that these free email servers proved quite inconsistent. If there is a way that I can check/receive the emails generated from the system without using any thrid party interference it would be great

benpatterson commented:
@kashifch Could you enter an IT ticket to create test users for you with email addresses at edX that you can control? It'll just be a bit better to make the test users something that can be centrally maintained.

kashifch commented:
@benpatterson I like the idea, I think we would need just one user, how about e2etests@edx.org, we can just use aliases to derive more users

benpatterson commented:
Great. Could you put in an IT ticket please?

Muddasser commented:
Can one of the admins verify this patch?

kashifch commented:
@benpatterson, @raeeschachar, @Muddasser 

Please review this branch, I have removed the email client part so it is just the api client for now. I have incorporated al the feedback as well. Once it is merged I will create a new branch for the next chunk of code

feanil commented:
Closing due to age and lack of activity.
